mne_sandbox/connectivity/cfc.py

```diff
@@ -0,0 +1,506 @@
+        print('Some events will be cut off!')
+        ev = ev[msk_keep]
+
+    # Pull events from the raw data
+    epochs = []
+    for ix in ev:
+        ix_min, ix_max = [ix + int(i_tlim * sfreq)
+                          for i_tlim in [tmin, tmax]]
+        epochs.append(x[np.newaxis, :, ix_min:ix_max])
+    epochs = np.concatenate(epochs, axis=0)
+    times = np.arange(epochs.shape[-1]) / float(sfreq) + tmin
+    return epochs, times, msk_keep
+
+
+# For the viz functions
+def _extract_phase_and_amp(data_phase, data_amp, sfreq, freqs_phase,
+                           freqs_amp, scale=True):
+    """Extract the phase and amplitude of two signals for PAC viz.
+    data should be shape (n_epochs, n_times)"""
+    from sklearn.preprocessing import scale
+
+    # Morlet transform to get complex representation
+    band_ph = cwt_morlet(data_phase, sfreq, freqs_phase)
+    band_amp = cwt_morlet(data_amp, sfreq, freqs_amp)
+
+    # Calculate the phase/amplitude of relevant signals across epochs
+    band_ph_stacked = np.hstack(np.real(band_ph))
+    angle_ph = np.hstack(np.angle(band_ph))
+    amp = np.hstack(np.abs(band_amp) ** 2)
+
+    # Scale the amplitude for viz so low freqs don't dominate highs
+    if scale is True:
+        amp = scale(amp, axis=1)
+    return angle_ph, band_ph_stacked, amp
+
+
+def _pull_data(inst, ix_ph, ix_amp, ev=None, tmin=None, tmax=None):
+    """Pull data from either Base or Epochs instances"""
+    from mne.io.base import _BaseRaw
+    from mne.epochs import _BaseEpochs
+    if isinstance(inst, _BaseEpochs):
+        data_ph = inst._data[:, ix_ph, :]
+        data_amp = inst._data[:, ix_amp, :]
+    elif isinstance(inst, _BaseRaw):
+        data = inst[[ix_ph, ix_amp], :][0].squeeze()
+        data_ph, data_amp = [i[np.newaxis, ...] for i in data]
+    return data_ph, data_amp
+
+
+def phase_locked_amplitude(inst, freqs_phase, freqs_amp,
+                           ix_ph, ix_amp, tmin=-.5, tmax=.5,
+                           mask_times=None):
+    """Calculate the average amplitude of a signal at a phase of another.
+
+    Parameters
+    ----------
+    inst : instance of mne.Epochs or mne.io.Raw
+        The data to be used in phase locking computation
+    freqs_phase : np.array
+        The frequencies to use in phase calculation. The phase of each
+        frequency will be averaged together.
+    freqs_amp : np.array
+        The frequencies to use in amplitude calculation.
+    ix_ph : int
+        The index of the signal to be used for phase calculation
+    ix_amp : int
+        The index of the signal to be used for amplitude calculation
+    tmin : float
+        The time to include before each phase peak
+    tmax : float
+        The time to include after each phase peak
+    mask_times : np.array, dtype bool, shape (inst.n_times,)
+        If inst is an instance of Raw, this will only include times contained
+        in mask_times.
+
+    Returns
+    -------
+    data_amp : np.array
```

Comment:
(this one is a little awkward because there's already a reference to `n_times` earlier, so I'm calling it `n_times_window`)